### PR TITLE
Implement TCP keepalive support from RFC 1122

### LINKED ISSUE #34044: TCP connection does not support keepalive function
The keepalive function in tcp is very important. When will Zephyr OS support this function?

### PR DESCRIPTION
**Pull request by maweicoltsmart**
Fixes zephyrproject-rtos#34044
 Signed-off-by:  Horse Ma <mawei@coltsmart.com>

### CODE REVIEW

File: samples/net/sockets/echo_client/src/tcp.c [hunk 1]
```diff
@@ -158,7 +158,33 @@ static int start_tcp_proto(struct data *data, struct sockaddr *addr,
 		ret = -errno;
 	}
 #endif
-
+#if defined(CONFIG_NET_TCP_KEEPALIVE)
+	int keep_alive = true; // enable keep alive probe
+	int keep_idle = 5; // 5 Second
+	int keep_intvl = 5; // 5 Second
+	int keep_cnt = 3; // probe 3 times
```
Comment (jukkar) [inline samples/net/sockets/echo_client/src/tcp.c:165]: Please use `/* */` comments

File: subsys/net/ip/tcp.c [hunk 2]
```diff
@@ -545,6 +545,9 @@ static void tcp_send_timer_cancel(struct tcp *conn)
 	}
 
 	k_work_cancel_delayable(&conn->send_timer);
+#if defined(CONFIG_NET_TCP_KEEPALIVE)
+	k_work_cancel_delayable(&conn->keepalive_timer);
+#endif /*CONFIG_NET_TCP_KEEPALIVE */
```
Comment (jukkar) [inline subsys/net/ip/tcp.c:550]: style nit, this looks better if `/* CONFIG....`, please do search and replace for the whole file

File: subsys/net/ip/tcp.c [hunk 3]
```diff
@@ -879,6 +882,132 @@ static int get_tcp_nodelay(struct tcp *conn, void *value, size_t *len)
...
+static int set_tcp_keep_alive(struct tcp *conn, const void *value, size_t len)
+{
+	int keep_alive;
+
+	if (len != sizeof(int)) {
+		return -EINVAL;
+	}
+
+	keep_alive = *(int *)value;
+
```
Comment (jukkar) [inline subsys/net/ip/tcp.c:895]: we typically do not put empty line between assignment and check

File: subsys/net/ip/tcp.c [hunk 4]
```diff
@@ -879,6 +882,132 @@ static int get_tcp_nodelay(struct tcp *conn, void *value, size_t *len)
...
+static int set_tcp_keep_idle(struct tcp *conn, const void *value, size_t len)
+{
+	int keep_idle;
+
+	if (len != sizeof(int)) {
+		return -EINVAL;
+	}
+
+	keep_idle = *(int *)value;
+
```
Comment (jukkar) [inline subsys/net/ip/tcp.c:914]: ditto

File: subsys/net/ip/tcp.c [hunk 5]
```diff
@@ -879,6 +882,132 @@ static int get_tcp_nodelay(struct tcp *conn, void *value, size_t *len)
...
+static int set_tcp_keep_intvl(struct tcp *conn, const void *value, size_t len)
+{
+	int keep_intvl;
+
+	if (len != sizeof(int)) {
+		return -EINVAL;
+	}
+
+	keep_intvl = *(int *)value;
+
```
Comment (jukkar) [inline subsys/net/ip/tcp.c:933]: ditto

File: subsys/net/ip/tcp.c [hunk 6]
```diff
@@ -879,6 +882,132 @@ static int get_tcp_nodelay(struct tcp *conn, void *value, size_t *len)
...
+static int set_tcp_keep_cnt(struct tcp *conn, const void *value, size_t len)
+{
+	int keep_cnt;
+
+	if (len != sizeof(int)) {
+		return -EINVAL;
+	}
+
+	keep_cnt = *(int *)value;
+
```
Comment (jukkar) [inline subsys/net/ip/tcp.c:952]: ditto

File: subsys/net/ip/tcp.c [hunk 7]
```diff
@@ -879,6 +882,132 @@ static int get_tcp_nodelay(struct tcp *conn, void *value, size_t *len)
...
+static int get_tcp_keep_alive(struct tcp *conn, void *value, size_t *len)
+{
+	int keep_alive = (int)conn->keep_alive;
+
+	*((int *)value) = keep_alive;
+
+	if (len) {
+		*len = sizeof(int);
+	}
+	return 0;
```
Comment (jukkar) [inline subsys/net/ip/tcp.c:971]: also we usually put empty line between `}` and next statement

File: subsys/net/ip/tcp.c [hunk 8]
```diff
@@ -879,6 +882,132 @@ static int get_tcp_nodelay(struct tcp *conn, void *value, size_t *len)
...
+static int get_tcp_keep_idle(struct tcp *conn, void *value, size_t *len)
+{
+	int keep_idle = (int)conn->keep_idle;
+
+	*((int *)value) = keep_idle;
+
+	if (len) {
+		*len = sizeof(int);
+	}
+	return 0;
```
Comment (jukkar) [inline subsys/net/ip/tcp.c:983]: ditto

File: subsys/net/ip/tcp.c [hunk 9]
```diff
@@ -879,6 +882,132 @@ static int get_tcp_nodelay(struct tcp *conn, void *value, size_t *len)
...
+static int get_tcp_keep_intvl(struct tcp *conn, void *value, size_t *len)
+{
+	int keep_intvl = (int)conn->keep_intvl;
+
+	*((int *)value) = keep_intvl;
+
+	if (len) {
+		*len = sizeof(int);
+	}
+	return 0;
```
Comment (jukkar) [inline subsys/net/ip/tcp.c:995]: and here

File: subsys/net/ip/tcp.c [hunk 10]
```diff
@@ -879,6 +882,132 @@ static int get_tcp_nodelay(struct tcp *conn, void *value, size_t *len)
...
+static int get_tcp_keep_cnt(struct tcp *conn, void *value, size_t *len)
+{
+	int keep_cnt = (int)conn->keep_cnt;
+
+	*((int *)value) = keep_cnt;
+
+	if (len) {
+		*len = sizeof(int);
+	}
+	return 0;
```
Comment (jukkar) [inline subsys/net/ip/tcp.c:1007]: and here too

File: subsys/net/ip/tcp.c [hunk 11]
```diff
@@ -1283,6 +1414,76 @@ static void tcp_fin_timeout(struct k_work *work)
...
+	}
+
+	NET_DBG("conn: %p keepalive probe", conn);
+	k_work_reschedule_for_queue(
+		&tcp_work_q, &conn->keepalive_timer,
+		K_SECONDS(conn->keep_intvl));
+
+	pkt = tcp_pkt_alloc(conn, sizeof(struct tcphdr));
+	if (!pkt) {
+		ret = -ENOBUFS;
```
Comment (jukkar) [inline subsys/net/ip/tcp.c:1447]: this is no op as the function does not return anything

File: subsys/net/ip/tcp.c [hunk 12]
```diff
@@ -1283,6 +1414,76 @@ static void tcp_fin_timeout(struct k_work *work)
...
+		goto out;
+	}
+
+	ret = ip_header_add(conn, pkt);
+	if (ret < 0) {
+		tcp_pkt_unref(pkt);
+		goto out;
```
Comment (jukkar) [inline subsys/net/ip/tcp.c:1454]: perhaps nicer to add `unref:` label at the end and jump to it

File: subsys/net/ip/tcp.c [hunk 13]
```diff
@@ -1283,6 +1414,76 @@ static void tcp_fin_timeout(struct k_work *work)
...
+	}
+
+	ret = tcp_header_add(conn, pkt, ACK, conn->seq - 1);
+	if (ret < 0) {
+		tcp_pkt_unref(pkt);
+		goto out;
```
Comment (jukkar) [inline subsys/net/ip/tcp.c:1460]: and here

File: subsys/net/ip/tcp.c [hunk 14]
```diff
@@ -1369,6 +1570,15 @@ static struct tcp *tcp_conn_alloc(struct net_context *context)
 	k_work_init_delayable(&conn->recv_queue_timer, tcp_cleanup_recv_queue);
 	k_work_init_delayable(&conn->persist_timer, tcp_send_zwp);
 
+#if defined(CONFIG_NET_TCP_KEEPALIVE)
+	conn->keep_alive = false;
+	conn->keep_idle = CONFIG_NET_TCP_KEEPIDLE_DEFAULT;
+	conn->keep_intvl = CONFIG_NET_TCP_KEEPINTVL_DEFAULT;
+	conn->keep_cnt = CONFIG_NET_TCP_KEEPCNT_DEFAULT;
+	NET_INFO("keepalive timer init idle = %d, interval = %d, cnt = %d", conn->keep_idle, conn->keep_intvl, conn->keep_cnt);
```
Comment (jukkar) [inline subsys/net/ip/tcp.c:1578]: Not sure if this needs to be in info level, would user be interested to see this always. I would suggest `NET_DBG` here instead.

**Review by jukkar — COMMENTED** [review]
Looks good, minor style nits.

**Comment by ssharks** [comment]
@maweicoltsmart Although the review process is intended to aid developers, I can imagine you are overwhelmed by the amount of review comments. 

Some tips you can used to check by yourself:

Run the most prominent tests for TCP:
./scripts/twister -T tests/net/socket/tcp -p qemu_x86
./scripts/twister -T tests/net/tcp -p qemu_x86

Check your head commit before pushing against a lot of style rules
./scripts/checkpatch.pl -g HEAD

I hope this helps you with this and hopefully many more PRs

File: subsys/net/ip/tcp.c [hunk 15]
```diff
@@ -545,6 +545,9 @@ static void tcp_send_timer_cancel(struct tcp *conn)
 	}
 
 	k_work_cancel_delayable(&conn->send_timer);
+#if defined(CONFIG_NET_TCP_KEEPALIVE)
+	k_work_cancel_delayable(&conn->keepalive_timer);
```
[the remark this answers is not in the document]
Comment (maweicoltsmart) [reply subsys/net/ip/tcp.c:552]: Yes, this code is redundant and may cause keepalive to fail, I'll correct it right away.

**Pull request closed without merging**